Add support for Aion [chain integration]

ivan-trust opened the pull request:
## Description

Added support for Aion.

Updated HDWallet, PrivateKey and PublicKey to support Aion key format.

## Checklist

- [x] Add tests to cover changes as needed.
- [x] Return correct curve and purpose in `src/Coin.cpp`.
- [x] Derivation from private key in `src/Coin.cpp`.
- [x] Validation in `src/Coin.cpp`.

hewigovens commented on src/Aion/Transaction.cpp:
```diff
@@ -0,0 +1,30 @@
+    append(encoded, Ethereum::RLP::encode(to.bytes));
+    append(encoded, Ethereum::RLP::encode(amount));
+    append(encoded, Ethereum::RLP::encode(payload));
+    append(encoded, Ethereum::RLP::encode(uint128_t(155157377101))); // Huge timestamp
```
what's this hardcoded value?

ivan-trust replied:
Aion-web3 takes [current timestamp * 1000](https://github.com/aionnetwork/aion_web3/blob/52453c82aa353f74f66f17f67d7d1f796c238d99/packages/web3-eth-accounts/src/index.js#L211) when building a transaction, if it's not provided by user. I just hardcoded some huge timestamp.

hewigovens commented on src/HDWallet.cpp:
```diff
@@ -146,22 +147,46 @@ std::optional<std::string> HDWallet::getAddressFromExtended(const std::string& e
+        
+        auto correctedChange = change;
+        auto correctedAddress = address;
+        if (coin == TWCoinTypeAion) {
```
do `switch` here as well?

vikmeup replied:
I would find a way to not have a switch for coins in this function, @alejandro-isaza might have more insight on this

hewigovens replied:
we may not need change / address for curve ed25519 this? https://github.com/satoshilabs/slips/blob/master/slip-0010.md

ivan-trust replied:
I did `switch` for now.

hewigovens commented:
could you please rebase master and fix CI (Failed - Wallet Core) ? it would be nice to have swift interface tests (test AionAddress.swift and AionSigner.swift)

terorie commented on include/TrustWalletCore/TWPublicKey.h:
```diff
@@ -38,6 +41,9 @@ TWData *_Nonnull TWPublicKeyData(struct TWPublicKey pk);
 bool TWPublicKeyVerify(struct TWPublicKey pk, TWData *_Nonnull signature, TWData *_Nonnull message);
 
+TW_EXPORT_METHOD
+bool TWPublicKeyVerifyEd25519(struct TWPublicKey pk, TWData *_Nonnull signature, TWData *_Nonnull message);
```
https://github.com/TrustWallet/wallet-core/pull/76
would expose a nicer C and C++ API

ivan-trust replied:
Great, I'll update my PR.

terorie replied:
@ivan-trust Please don't, they were just reverted 😬 
@alejandro-isaza Is working on a different API, which will be released soon.

ivan-trust commented:
> could you please rebase master and fix CI (Failed - Wallet Core) ? it would be nice to have swift interface tests (test AionAddress.swift and AionSigner.swift)

Rebased master, added Swift tests.

vikmeup commented on swift/Sources/Blockchains/Aion.swift:
```diff
@@ -0,0 +1,42 @@
+extension AionAddress: Address {
+    
+    public static func isValid(data: Data) -> Bool {
+        if let _ = AionAddress(data: data) {
```
`return AionAddress(data: data) != nil`

hewigovens commented:
overall looks good, Is there any other HD wallets support Aion? so we can cross verify the addresses derived from phrases.

alejandro-isaza commented:
Updated checklist, please complete.

ivan-trust commented:
> Updated checklist, please complete.
> https://github.com/TrustWallet/wallet-core/blob/master/.github/ISSUE_TEMPLATE/new_blockchain.md#checklist

✅
